pkg/utils/fsutils/license_scanning_fs.go

```diff
@@ -0,0 +1,295 @@
+	mutex       sync.Mutex
+}
+
+type RecursiveWalkerInput struct {
+	Logger                    *slog.Logger
+	Parser                    types.PackageManifestParser
+	PackageManifestFile       string
+	PackageDependencyDir      string
+	ClassifierConfidenceLevel float64
+	LicenseTextCacheDir       string
+	ParallelWorkers           int
+}
+
+// default constructor for Recursive walker
+// It also initializes the classifier pool within licensing package
+func NewRecursiveWalker(
+	input RecursiveWalkerInput,
+) (*RecursiveWalker, error) {
+	// Initialize the license classifier pool before the recursive walker
+	err := licensing.InitGoogleLicenseClassifierPool(input.ParallelWorkers)
+	if err != nil {
+		return nil, err
+	}
+
+	return &RecursiveWalker{
+		RecursiveWalkerInput: input,
+		licenses:             make(map[string][]types.License),
+		pkgIDMap:             sync.Map{},
+		processChan:          make(chan licensing.ClassifierInput, 2*input.ParallelWorkers),
+	}, nil
+}
+
+// starts the worker threads based on given number of workers
+func (w *RecursiveWalker) StartWorkerPool() {
+	for i := 0; i < w.ParallelWorkers; i++ {
+		w.waitGroup.Add(1)
+		go w.StartWorker()
+	}
+}
+
+// stops the worker threads and closes their respective channels
+func (w *RecursiveWalker) StopWorkerPool() {
+	// close the process chan to stop all the worker go routines
+	close(w.processChan)
+
+	// wait for all threads to finish
+	w.waitGroup.Wait()
+}
+
+// starts individual worker thread. Lists on Process channel and sends data to license classifier
+func (w *RecursiveWalker) StartWorker() {
+	defer w.waitGroup.Done()
+
+	for classifierInput := range w.processChan {
+		pkgID := classifierInput.PkgID
+		concludedLicenses, err := checkForConcludedLicenses(classifierInput)
+		if err != nil {
+			w.Logger.Error("failed to get concluded licenses",
+				log.Any("input", classifierInput), log.String("error", err.Error()))
+			continue
+		}
+
+		if len(concludedLicenses) > 0 {
+			w.Logger.Debug("Found concluded licenses",
+				log.String("pkgID", pkgID), log.Any("concludedLicenses", concludedLicenses))
+
+			w.mutex.Lock()
+			w.licenses[pkgID] = append(w.licenses[pkgID], concludedLicenses...)
+			w.mutex.Unlock()
+		}
+	}
+}
+
+// Recursive walker walks the given fs and gets the concluded licenses.
+// It's Used for deep license scanning.
+// Note: For root as ".", we scan all the files and add them as loose licenses
+func (w *RecursiveWalker) Walk(fsys fs.FS, root string, parentPkgID string) (bool, error) {
+	if isSpecialPath(root) {
+		return w.handleSpecialPath(fsys, root)
+	}
+
+	pkgID, foundPkgManifest, err := w.processPackageManifest(fsys, root)
+	if err != nil {
+		w.Logger.Error("Failed to process package manifest", log.String("Err", err.Error()))
+		return false, err
+	}
+
+	if !foundPkgManifest {
+		if parentPkgID == "" {
+			w.Logger.Debug("Parent PkgID is empty. Adding to loose licenses", log.String("path", root))
+			pkgID = types.LOOSE_LICENSES
```

Comment:
i was under the impression that this was fixed. We discussed that using the same variable for 2 different types of values is confusing